chia/util/puzzle_compression.py

```diff
@@ -191,58 +191,81 @@ def solve(args: List[Union[PuzzleRepresentation, Program]], solution_dict: Dict[
-            raise CompressionVersionError()
+    version_number: uint16
+    driver_dict: Dict[bytes32, Any]
+
+    def __init__(self, version: uint16 = LATEST_VERSION, driver_dict: Dict[bytes32, Any] = None):
+        if driver_dict is None:
+            assert version in HASH_TO_DRIVER.keys()
+            self.version_number = uint16(version)
+            final_dict: Dict[bytes32, Any] = {}
+            for key in range(0, version + 1):
+                final_dict = final_dict | HASH_TO_DRIVER[uint16(key)]
+            self.driver_dict = final_dict
         else:
-            return object_bytes[2:]
+            self.version_number = uint16(65535)  # Just set it to max so it always passes any version checks
+            self.driver_dict = driver_dict
 
     @classmethod
-    def get_identifier_version(cls, identifier: bytes32) -> uint16:
+    def get_identifier_version(cls, identifier: bytes32) -> "CompressorVersion":
         for id, driver_dict in HASH_TO_DRIVER.items():
             if identifier in driver_dict:
-                return id
+                return cls(id)
         raise ValueError("The given identifier does not have a puzzle driver mapping")
 
     @classmethod
-    def identify_lowest_compatible_version(cls, matched_puzzles: List[Union[PuzzleRepresentation, Program]]) -> uint16:
-        highest_version: uint16 = uint16(0)
+    def lowest_compatible_version(
+        cls, matched_puzzles: List[Union[PuzzleRepresentation, Program]]
+    ) -> "CompressorVersion":
+        highest_version = uint16(0)
         for puz in matched_puzzles:
             if isinstance(puz, PuzzleRepresentation):
                 highest_version = uint16(
                     max(
                         highest_version,
-                        cls.get_identifier_version(puz.base),
-                        cls.identify_lowest_compatible_version(puz.args),
+                        cls.get_identifier_version(puz.base).version_number,
+                        cls.lowest_compatible_version(puz.args).version_number,
                     )
                 )
-        return highest_version
+        return cls(highest_version)
+
+
+class CompressionVersionError(Exception):
+    pass
+
+
+class PuzzleCompressor:
+    @classmethod
+    def match_puzzle(
+        cls, puzzle: Program, version=CompressorVersion()
+    ) -> Tuple[bool, Union[PuzzleRepresentation, Program]]:
+        for identifier, driver in version.driver_dict.items():
+            matched, args = driver.match(puzzle, version=version)
+            if matched:
+                return True, PuzzleRepresentation(identifier, args)
+        return False, puzzle
+
+    @classmethod
+    def serialize(cls, puzzle: Program, version=CompressorVersion()) -> bytes:
+        _, rep = cls.match_puzzle(puzzle, version=version)
+        return bytes(version.version_number) + bytes(rep)
+
+    @classmethod
+    def deserialize(cls, object_bytes: bytes, version=CompressorVersion()) -> Program:
```

Comment:
You don't need a version number for each sub-puzzle; just a version number at the start of the recursive compressed puzzle structure. You turn that into a specific `CompressorVersion` object (which is now probably incorrectly-named, since it doesn't need version information in it anymore), and pass that one object through the whole hierarchy of calls.